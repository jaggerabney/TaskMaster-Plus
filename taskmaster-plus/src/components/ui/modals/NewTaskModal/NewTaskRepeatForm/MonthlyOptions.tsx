import React from "react";

import {
  MonthlyOptionsType,
  daysOfTheWeek,
  setPosDropdownDict
} from "@/utils/Util";

interface MonthlyOptionsPropsType {
  data: MonthlyOptionsType;
  onChange: (data: MonthlyOptionsType) => void;
}

const MonthlyOptions: React.FC<MonthlyOptionsPropsType> = ({
  data,
  onChange
}) => {
  const { basis, byMonthDay, bySetPos, byDay } = data;

  function inputChangeHandler(
    event:
      | React.ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLSelectElement>
  ) {
    const targetComponentId = event.target.id;
    let newBasis: string = basis;
    let newByMonthDay: number = byMonthDay;
    let newBySetPos: number = bySetPos;
    let newByDay: string = byDay;

    switch (targetComponentId) {
      case "by-monthday-checkbox":
        newBasis = "BYMONTHDAY";
        break;
      case "by-setpos-checkbox":
        newBasis = "BYSETPOS";
        break;
      case "by-monthday-input":
        newByMonthDay = Number(event.target.value);
        break;
      case "by-setpos-dropdown":
        newBySetPos = Number(event.target.value);
        break;
      case "by-day-dropdown":
        newByDay = event.target.value.slice(0, 2).toUpperCase();
        break;
    }

    onChange({
      basis: newBasis,
      byMonthDay: newByMonthDay,
      bySetPos: newBySetPos,
      byDay: newByDay
    });
  }

  return (
    <ul className="flex flex-col text-xl gap-4">
      <li className="flex flex-row items-center gap-4 ml-8">
        <input
          id="by-monthday-checkbox"
          type="checkbox"
          onChange={inputChangeHandler}
          checked={basis === "BYMONTHDAY"}
        />
        <label htmlFor="by-monthday-checkbox">On day</label>
        <input
          id="by-monthday-input"
          onChange={inputChangeHandler}
          className="border border-black rounded p-2"
          type="number"
          step={1}
          min={1}
          max={31}
          value={byMonthDay}
        />
      </li>
      <li className="flex flex-row items-center gap-4 ml-8">
        <input
          id="by-setpos-checkbox"
          type="checkbox"
          onChange={inputChangeHandler}
          checked={basis === "BYSETPOS"}
        />
        <label htmlFor="by-setpos-checkbox">On the</label>
        <select
          id="by-setpos-dropdown"
          value={bySetPos}
          onChange={inputChangeHandler}
          className="border border-black rounded p-2"
        >
          {Array.from(setPosDropdownDict.keys()).map((key) => (
            <option key={key} value={setPosDropdownDict.get(key)}>
              {key}
            </option>
          ))}
        </select>
        <select
          id="by-day-dropdown"
          value={daysOfTheWeek.find((day) =>
            day.toUpperCase().startsWith(byDay)
          )}
          onChange={inputChangeHandler}
          className="border border-black rounded p-2"
        >
          {daysOfTheWeek.map((day) => (
            <option key={day} value={day}>
              {day}
            </option>
          ))}
        </select>
      </li>
    </ul>
  );
};

export default MonthlyOptions;
