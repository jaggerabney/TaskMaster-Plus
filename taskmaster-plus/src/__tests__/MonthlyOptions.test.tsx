import "@testing-library/jest-dom";
import { render, fireEvent, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import MonthlyOptions from "@/components/ui/modals/NewTaskModal/NewTaskRepeatForm/MonthlyOptions";
import {
  defaultRepeatFormState,
  setPosDropdownDict,
  MonthlyOptionsType,
  daysOfTheWeek
} from "@/utils/Util";

describe("Monthly options component", () => {
  const mockOnChange = jest.fn();
  const mockMonthlyOptions = (
    <MonthlyOptions
      data={defaultRepeatFormState.monthly}
      onChange={mockOnChange}
    />
  );
  let mockOnChangeCalls = 0;

  it("should render correctly", () => {
    const monthlyOptions = render(mockMonthlyOptions);

    expect(monthlyOptions).toMatchSnapshot();
  });

  // basis checkboxes

  it("allows user to change basis", async () => {
    render(mockMonthlyOptions);
    const byMonthDayCheckbox = document.getElementById(
      "by-monthday-checkbox"
    ) as HTMLInputElement;
    const bySetPosCheckbox = document.getElementById(
      "by-setpos-checkbox"
    ) as HTMLInputElement;
    let onChangeData: MonthlyOptionsType;

    await userEvent.click(bySetPosCheckbox);
    onChangeData = mockOnChange.mock.calls[mockOnChangeCalls++][0];
    expect(onChangeData.basis).toEqual("BYSETPOS");

    await userEvent.click(byMonthDayCheckbox);
    onChangeData = mockOnChange.mock.calls[mockOnChangeCalls++][0];
    expect(onChangeData.basis).toEqual("BYMONTHDAY");
  });

  it('sets the basis to "On day" by default', () => {
    render(mockMonthlyOptions);
    const byMonthDayCheckbox = document.getElementById(
      "by-monthday-checkbox"
    ) as HTMLInputElement;

    expect(byMonthDayCheckbox.checked).toBe(true);
  });

  // BYMONTHDAY input

  it("allows user to change BYMONTHDAY input", () => {
    render(mockMonthlyOptions);
    const byMonthDayInput = document.getElementById(
      "by-monthday-input"
    ) as HTMLInputElement;

    fireEvent.change(byMonthDayInput, { target: { value: 3 } });

    const onChangeData: MonthlyOptionsType =
      mockOnChange.mock.calls[mockOnChangeCalls++][0];
    expect(onChangeData.byMonthDay).toBe(3);
  });

  it("sets BYMONTHDAY to 1 by default", () => {
    render(mockMonthlyOptions);
    const byMonthDayInput = document.getElementById(
      "by-monthday-input"
    ) as HTMLInputElement;

    expect(byMonthDayInput.value).toEqual("1");
  });

  // BYSETPOS dropdown

  it("allows user to change BYSETPOS dropdown", async () => {
    render(mockMonthlyOptions);
    const bySetPosDropdown = document.getElementById(
      "by-setpos-dropdown"
    ) as HTMLSelectElement;

    await userEvent.selectOptions(
      bySetPosDropdown,
      screen.getByRole("option", { name: "4th" })
    );

    const onChangeData: MonthlyOptionsType =
      mockOnChange.mock.calls[mockOnChangeCalls++][0];
    expect(onChangeData.bySetPos).toEqual(4);
  });

  it("only has five BYSETPOS options: 1st, 2nd, 3rd, 4th, and Last", () => {
    render(mockMonthlyOptions);
    const bySetPosDropdown = document.getElementById(
      "by-setpos-dropdown"
    ) as HTMLSelectElement;
    const setPosDropdownDictKeys = setPosDropdownDict.keys().toArray();
    const bySetPosDropdownOptions = [];

    for (const child of bySetPosDropdown.children) {
      bySetPosDropdownOptions.push(child.textContent);
    }

    expect(setPosDropdownDictKeys).toEqual(bySetPosDropdownOptions);
  });

  it('displays "1st" as the default BYSETPOS value', () => {
    render(mockMonthlyOptions);
    const bySetPosDropdown = document.getElementById(
      "by-setpos-dropdown"
    ) as HTMLSelectElement;

    expect(
      bySetPosDropdown.options[bySetPosDropdown.selectedIndex].text
    ).toEqual("1st");
  });

  // BYDAY dropdown

  it("allows user to change BYDAY dropdown", async () => {
    render(mockMonthlyOptions);
    const byDayDropdown = document.getElementById(
      "by-day-dropdown"
    ) as HTMLSelectElement;

    await userEvent.selectOptions(
      byDayDropdown,
      screen.getByRole("option", { name: "Friday" })
    );

    const onChangeData: MonthlyOptionsType =
      mockOnChange.mock.calls[mockOnChangeCalls++][0];
    expect(onChangeData.byDay).toEqual("FR");
  });

  it("only has seven BYDAY options - one for each day of the week", () => {
    render(mockMonthlyOptions);
    const byDayDropdown = document.getElementById(
      "by-day-dropdown"
    ) as HTMLSelectElement;
    const byDayDropdownOptions = [];

    for (const child of byDayDropdown.children) {
      byDayDropdownOptions.push(child.textContent);
    }

    expect(daysOfTheWeek).toEqual(byDayDropdownOptions);
  });

  it('displays "Sunday" as the default BYDAY value', () => {
    render(mockMonthlyOptions);
    const byDayDropdown = document.getElementById(
      "by-day-dropdown"
    ) as HTMLSelectElement;

    expect(byDayDropdown.options[byDayDropdown.selectedIndex].text).toEqual(
      "Sunday"
    );
  });
});
