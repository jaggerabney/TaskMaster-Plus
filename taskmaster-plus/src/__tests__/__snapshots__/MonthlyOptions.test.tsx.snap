// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monthly options component should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="flex flex-col text-xl gap-4"
        id="monthly-options"
      >
        <li
          class="flex flex-row items-center gap-4 ml-8"
        >
          <input
            checked=""
            id="by-monthday-checkbox"
            type="checkbox"
          />
          <label
            for="by-monthday-checkbox"
          >
            On day
          </label>
          <input
            class="border border-black rounded p-2"
            id="by-monthday-input"
            max="31"
            min="1"
            step="1"
            type="number"
            value="1"
          />
        </li>
        <li
          class="flex flex-row items-center gap-4 ml-8"
        >
          <input
            id="by-setpos-checkbox"
            type="checkbox"
          />
          <label
            for="by-setpos-checkbox"
          >
            On the
          </label>
          <select
            class="border border-black rounded p-2"
            id="by-setpos-dropdown"
          >
            <option
              value="1"
            >
              1st
            </option>
            <option
              value="2"
            >
              2nd
            </option>
            <option
              value="3"
            >
              3rd
            </option>
            <option
              value="4"
            >
              4th
            </option>
            <option
              value="-1"
            >
              Last
            </option>
          </select>
          <select
            class="border border-black rounded p-2"
            id="by-day-dropdown"
          >
            <option
              value="Sunday"
            >
              Sunday
            </option>
            <option
              value="Monday"
            >
              Monday
            </option>
            <option
              value="Tuesday"
            >
              Tuesday
            </option>
            <option
              value="Wednesday"
            >
              Wednesday
            </option>
            <option
              value="Thursday"
            >
              Thursday
            </option>
            <option
              value="Friday"
            >
              Friday
            </option>
            <option
              value="Saturday"
            >
              Saturday
            </option>
          </select>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="flex flex-col text-xl gap-4"
      id="monthly-options"
    >
      <li
        class="flex flex-row items-center gap-4 ml-8"
      >
        <input
          checked=""
          id="by-monthday-checkbox"
          type="checkbox"
        />
        <label
          for="by-monthday-checkbox"
        >
          On day
        </label>
        <input
          class="border border-black rounded p-2"
          id="by-monthday-input"
          max="31"
          min="1"
          step="1"
          type="number"
          value="1"
        />
      </li>
      <li
        class="flex flex-row items-center gap-4 ml-8"
      >
        <input
          id="by-setpos-checkbox"
          type="checkbox"
        />
        <label
          for="by-setpos-checkbox"
        >
          On the
        </label>
        <select
          class="border border-black rounded p-2"
          id="by-setpos-dropdown"
        >
          <option
            value="1"
          >
            1st
          </option>
          <option
            value="2"
          >
            2nd
          </option>
          <option
            value="3"
          >
            3rd
          </option>
          <option
            value="4"
          >
            4th
          </option>
          <option
            value="-1"
          >
            Last
          </option>
        </select>
        <select
          class="border border-black rounded p-2"
          id="by-day-dropdown"
        >
          <option
            value="Sunday"
          >
            Sunday
          </option>
          <option
            value="Monday"
          >
            Monday
          </option>
          <option
            value="Tuesday"
          >
            Tuesday
          </option>
          <option
            value="Wednesday"
          >
            Wednesday
          </option>
          <option
            value="Thursday"
          >
            Thursday
          </option>
          <option
            value="Friday"
          >
            Friday
          </option>
          <option
            value="Saturday"
          >
            Saturday
          </option>
        </select>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
